### ISSUE TITLE
Data fetching & virtualization

### ISSUE THREAD

**Issue by CHE1RON** [lukasbach/react-complex-tree#321]
Hello there,

first of all, let me say that I really enjoy working with this library, especially its focus on a11y is appreciated! Now, after working with uncontrolled tree data, I wondered if anyone has managed the tree data using a controlled environment powered by data fetching libraries like `react-query`?

Also I wondered if it's suitable for virtualization (as this works best with flat lists which does not work well with a11y and semantic tree structures, like nested lists etc) .. ?

I'm grateful for any pointers or documentation you might be able to provide!

Cheers
CHE1RON

**Comment by lukasbach** [comment]
Thanks for your appreciation! For your question regarding managing tree data with fetching libraries, I can't provide examples with that, but the use case should also work wikth uncontrolled environments which might be a bit easier to start with. I expanded the [docs on static data providers](https://rct.lukasbach.com/docs/guides/static-data-provider/) and on [custom data providers](https://rct.lukasbach.com/docs/guides/custom-data-provider/), maybe the examples there are helpful, the same way how in the docs items are injected, you could also alter the local tree data when something changes in the queried data. Of course it should also work with controlled environments, though you need to do a bit more heavy lifting there.
For your comment on virtualization, yes, that is a bit tricky with the nested tree data, maybe the comments in the thread you found can help. And yes, once Headless Tree is production ready, that would fill the same use case as RCT but with some additional capabilities and support for virtualization.

**Comment by Nishchit14** [comment]
@lukasbach I am eagerly waiting for headless-tree. Any CTA for it? Really appreciate roadmap clarity on it. Thanks for all your amazing OSS work.

**Comment by lukasbach** [comment]
Hey @Nishchit14, sorry for being silent on that topic and not bringing a lot of progress to that library. I had lots of topics on my mind the last few months, and wasn't really able to do much progress. I can't provide a specific timeline at the moment, but do plan to pick the remaining topics up soon and get a version suitable for production out during this summer.

**Comment by max-dirac** [comment]
Hey, kind of curious as to why the react-complex-tree isn't supportive of virtualization.

**Comment by CHE1RON** [comment]
@max-dirac Well, we're still waiting for your PR 😛

**Comment by max-dirac** [comment]
I've done some investigation and the technique for virtualization as I'm aware would require that the tree be stored in a flat list as opposed to a nested object. I've implemented this at my company as a fully custom tree library with a dependency on tanstack virtualize. I know @lukasbach is attempting to address this with his new project. Virtualization according to my understanding of react-complex-tree and my knowledge of implementation techniques would likely require a hefty architectural change to its core data structure.
